### ISSUE TITLE
[1822 family] Autorouting does not use E train properly

### ISSUE THREAD

**Issue by perwestling** [tobymao/18xx#7982]
In a game of 1822MX I clicked on autorouting when I was going to run E train for first time.
6(+) train had a route, the P3/2 had a route, and autorouting filled in the E train.

![image](https://user-images.githubusercontent.com/2631151/174977959-77e30bb8-976c-420c-b1b2-ad6dde892b4c.png)
https://18xx.games/game/87015?action=1254

But the chosen route (G22-F21-F23) is not a legal E route as F23 has no token!

And auto routing for E train does not seem to run on the same track as the other trains - which is allowed.

So this is an example where auto routing really misses the target.

Of course getting the right route when adding an E train is not difficult so I will manage.

**Comment by ChrisTheCat** [comment]
I have the same experience in this game https://18xx.games/game/123526?action=1145

**Comment by benjaminxscott** [comment]
@michaeljb did your prior commit fix this issue? (the "closes #XXX" was in the commit desc so it didn't get auto-applied)

**Comment by michaeljb** [comment]
> @michaeljb did your prior commit fix this issue? (the "closes #XXX" was in the commit desc so it didn't get auto-applied)
>  
> @benjaminxscott no, I didn't quite get that working. That commit is only on a branch on my fork, never merged into the main code.

Got it working now, opened a PR for it 🎉

**Comment by michaeljb** [comment]
My work on this caused #11036.  #11745 solved that, but has reintroduced this problem. @PistonPercy

**Comment by PistonPercy** [comment]
I’m away from a computer I can program from for a few days but can poke after. To be clear, my pull request was suppose to allow this to work (and built off your technique), but I did have annoying off by one errors + opal/JS boundary bugs that tripped me up, so it’s believable I messed up in more ways.

Do you have a specific example which doesn’t work that I’ll use to debug?

**Comment by michaeljb** [comment]
https://18xx.games/game/168857?action=1239 (this game is within like a week of being archived, so [here's the JSON in a gist as well](https://gist.github.com/michaeljb/9dceb050f2ead587d4e6e9fa144f3a7b))

I did just now get the best route for the E and the 5, when I hit the `Auto` button with a route only for the LP.

If one of the E or 5 has a route already when pressing `Auto`, the autorouter will avoid overlapping with the other--that's what I saw earlier.

**Comment by PistonPercy** [comment]
Ah I understand the bug! To clarify it, we won't re-use track if a pre-filled train has used it, but we will return overlapping routes of non-pre-filled trains. I'll try to get a fix asap.

In case it's habit based on stale info, I'd love to know the reason to be pre-filling routes, because in this case, clearing them all gives a better combo than the one that was selected in the game. I'd love to get a map where it's worth pre-filling a train because it significantly changes the routing time. 

My sense is it used to be very valuable to do after https://github.com/tobymao/18xx/pull/10929 and before https://github.com/tobymao/18xx/pull/11745 because the first patch gave each routing group it's own pathing timeout and each routing group did it's own graph traversal, but the second patch globbed those together. On a local repo, I've got a patch which will speed that up tremendously, so It's probably not worth debating which way is better. (I've seen 200x on a specific turn, but I'm not going claim that generally).

**Comment by ChrisTheCat** [comment]
The E train can overlap with regular train pre-filled routes, so I'm not sure the above is correct. And vice versa, if the E train is pre-filled, a regular train can overlap with the pre-filled E route.

**Comment by michaeljb** [comment]
> Ah I understand the bug! To clarify it, we won't re-use track if a pre-filled train has used it, but we will return overlapping routes of non-pre-filled trains. I'll try to get a fix asap.
>
> In case it's habit based on stale info, I'd love to know the reason to be pre-filling routes, because in this case, clearing them all gives a better combo than the one that was selected in the game. I'd love to get a map where it's worth pre-filling a train because it significantly changes the routing time.

The example I linked in my previous comment takes 27 seconds to auto route with nothing pre-filled. I was able to pre-fill the LP and E-train (whose routes are trivial here) with about 13 seconds of clicking, and then the autorouter takes about 2 seconds to pick a route for the 5-train. Sounds like your local changes might bring the 27 seconds down close enough to the 2?

More broadly speaking, as a user I expected it to route the train with stuff pre-filled because that's how it's worked for a long time on the site.

For another bad performance case, [this gist](https://gist.github.com/michaeljb/3d738a10ccf9965a42f91b6134253b0c) has the data for the game I was looking at in #10331. I don't know if that graph slowness actually does directly translate to autorouter slowness but that's the idea that came to mind.

> The E train can overlap with regular train pre-filled routes, so I'm not sure the above is correct. And vice versa, if the E train is pre-filled, a regular train can overlap with the pre-filled E route.

@ChrisTheCat not sure what you're trying to say here. The bug here is that the current behavior of the "Auto" button is that it won't let an E and a regular train overlap if one of them is pre-filled, even though--as you say--it should be able to.

**Comment by PistonPercy** [comment]
> The example I linked in my previous comment takes 27 seconds to auto route with nothing pre-filled. I was able to pre-fill the LP and E-train (whose routes are trivial here) with about 13 seconds of clicking, and then the autorouter takes about 2 seconds to pick a route for the 5-train.

Yeah but I'll note in this case, it's a bug that it is that fast. It's that fast because the 5-train isn't allowed to use anything path the LP + E train are using which trims the search significantly. That being said, it's a fair point. 

> More broadly speaking, as a user I expected it to route the train with stuff pre-filled because that's how it's worked for a long time on the site.


I don't think I said anything against that right? I'm just saying I'm not very convinced you as a human needs to do it.

>  Sounds like your local changes might bring the 27 seconds down close enough to the 2?

I was seeing 6-7 seconds with nothing prefilled, but I don't want to claim anything until all the bugs are fixed. The tacky commit is [here](https://github.com/PistonPercy/18xx/commit/954686e9edd49f9812f84ba52fc122126291e196). This is written off the cuff, but if I remember right, the main insights I noticed are that the:
- node.walk yields for every path that is added, but this is a ton of invalid routes because we only care about path chains with nodes on the end.
- you can cache at a node level the chains that it connects to another nodes with. This trims an immense amount of function calls
- If you are analyzing performance, the path search is basically looking for the longest path that doesn't throw ReusesCity or RouteTooLong. In the games I was analyzing that means "NoToken" and "wrong track type" errors were the vast majority of invalid routes. So there are two tricks I used for that, (1) make it possible to differentiate GameErrors that "routes can't grow out of" vs "routes can grow and fix the game error" (I really need a good name for this difference though) (2) Instead looking for all paths that start from each reachable hex, look for routes expanding out from the tokens, so my experimental path'ing will never find a route that is invalid because it's missing a token.
- And the fun little trick is we filter out trains during the tree search once we know that train won't be valid for path extensions. 

> I don't know if that graph slowness actually does directly translate to autorouter slowness but that's the idea that came to mind.

It's probably possible.. the truth is I abandoned node.walk pretty quick in my diff because its interface has so many options I didn't understand and I really needed to constrain + cache the search. I wasn't clear if my chains caching could be ported back into node walk. It might be able, but maybe not because the autorouter is just so insanely hot loop that custom writing the 20 lines of code is worth it. Depending how long it takes for me to test that it is correct, I'll try to see if I can merge in caching.

I don't fully understand what you are computing, so feel free to ignore me. But something feels very wrong unless I'm misunderstanding what graph.compute is doing. If I'm understanding right, it is looking for all paths, hexes and nodes connected to a token and that should be O(n) with size of the graph. node.walk is more powerful than what graph.compute needs I think because graph.compute doesn't care about enumerating all path chains that exist from a node (the autorouter doesn't even care about that).

That being said https://github.com/PistonPercy/18xx/commit/65efce2103b33475f7f1c1123030d3f9df87ecee is my guess at the solution, but I didn't test it beyond loading that map (I really have to get back to fixing this bug and finalizing my pathing change in my 18xx coding time lol). Basically, before the walk would continue even if you already saw the path.. but with this change, if you have seen the path, don't keep walking down this route because presumably anything visited through this path is visited from something else? EDIT: https://github.com/PistonPercy/18xx/commit/684439d9833d66baa9eb52bb4b372efaef0f1343 fixes a bug because you need to take into account direction that the path is traveled. There has to be a better way to do it than the way I did it, but  I think it works.
